### PR TITLE
[CFP-326] Rake task to create Canarytokens on S3

### PR DESCRIPTION
**Pull request by jrmhaig**
#### What

Upload Canary documents to the S3 buckets for each CCCD environments with names that will attract a malicious user.

#### Ticket

[Add canaries to S3 storage](https://dsdmoj.atlassian.net/jira/software/c/projects/CFP/boards/624?modal=detail&selectedIssue=CFP-326)

#### Why

If an unauthorised user gains access to the S3 buckets then they may download and open files. If one of the Canarytokens is opened then we will get alerted to the breach.

#### How

A new Rake task will create Canarytokens and upload them to S3 as;

* `admin-users.docx`; Based on `docs/samples/canary_base.docx`, containing a table of fake users and passwords.
* `tmp/contents.pdf`;  Based on `docs/samples/canary_base.pdf` - a mostly blank file containing 'TBC'.

#### TODO

 - [x] Add some notes in `docs/*`
 - [x] Run the Rake task on dev-lgfs
 - [x] Run the Rake task on dev
 - [x] Run the Rake task on staging
 - [x] Run the Rake task on api-sandbox
 - [ ] Run the Rake task on production

### CODE REVIEW

File: spec/lib/thinkst_canary/token/shared_examples.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+          .to have_received(:query)
+          .with(:post, '/api/v1/canarytoken/factory/create', auth: false, params: request_options)
+      end
+
+      it { expect(token.canarytoken).to eq(canarytoken) }
+    end
+
+    context 'when using an existing Canary token' do
+      let(:existing_canarytoken) { 'existing_canarytoken' }
+      let(:token_options) { super().merge(canarytoken: existing_canarytoken) }
```
Comment (jsugarman) [inline spec/lib/thinkst_canary/token/shared_examples.rb:40]: never seen this. what is it it calling super on? rspec

File: spec/lib/thinkst_canary/token/shared_examples.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
+require 'rails_helper'
+
+RSpec.shared_examples 'a Canary token' do |kind|
+  describe '.new' do
+    subject(:token) { described_class.new(**token_options) }
+
+    let(:canarytoken) { 'canarytoken' }
```
Comment (jsugarman) [inline spec/lib/thinkst_canary/token/shared_examples.rb]: i think `new_canaryotoken` would be clearer here. relative to existing canarytoken (and the term is use a lot in theses test as well)

File: spec/lib/thinkst_canary/token/pdf_acrobat_reader_spec.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+require 'lib/thinkst_canary/token/shared_examples'
+
+RSpec.describe ThinkstCanary::Token::PdfAcrobatReader do
+  include_examples 'a Canary token', 'pdf-acrobat-reader' do
+    let(:file_upload) { instance_double(Faraday::UploadIO) }
+    let(:extra_token_options) { { file: StringIO.new } }
+    let(:extra_request_options) { { pdf: file_upload } }
+
+    before { allow(Faraday::UploadIO).to receive(:new).and_return(file_upload) }
```
Comment (jsugarman) [inline spec/lib/thinkst_canary/token/pdf_acrobat_reader_spec.rb]: do these belong here or in the shared example? They are repeated for the `doc-msword` version implying they could reside in the shared example - only the key `pdf:` is different

File: lib/thinkst_canary/token/file_based.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+module ThinkstCanary
+  module Token
+    class FileBased < Base
+      def initialize(**kwargs)
+        @file = kwargs[:file]
+
+        super(**kwargs)
```
Comment (jsugarman) [inline lib/thinkst_canary/token/file_based.rb]: just `super` needed to pass all args implicitly

File: lib/thinkst_canary/token/doc_msword.rb [hunk 5]
```diff
@@ -1,22 +1,12 @@
...
   module Token
-    class DocMsword < Base
+    class DocMsword < FileBased
       def initialize(**kwargs)
-        @kind = 'doc-msword'
-        @file = kwargs[:file]
+        @file_key = :doc
+        @mime = 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
 
         super(**kwargs)
```
Comment (jsugarman) [inline lib/thinkst_canary/token/doc_msword.rb]: just `super` needed to pass all args implicitly

File: lib/thinkst_canary/token/null_token.rb [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
   module Token
     class NullToken < Base
+      attr_reader :kind
+
       def initialize(**kwargs)
+        @kind = kwargs[:kind]
         @canarytoken = "Unknown Canary kind '#{kwargs[:kind]}'"
 
-        super
+        super(**kwargs)
```
Comment (jsugarman) [inline lib/thinkst_canary/token/null_token.rb]: just `super` needed to pass all args implicitly

File: lib/thinkst_canary/token/pdf_acrobat_reader.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+module ThinkstCanary
+  module Token
+    class PdfAcrobatReader < FileBased
+      def initialize(**kwargs)
+        @file_key = :pdf
+        @mime = 'application/pdf'
+
+        super(**kwargs)
```
Comment (jsugarman) [inline lib/thinkst_canary/token/pdf_acrobat_reader.rb]: just `super` needed to pass all args implicitly

**Review by jsugarman — CHANGES_REQUESTED** [review]
👌 Some test sharing issues and tweaks, plus a `super` issue to check

File: docs/thinkst_canary.md [hunk 8]
```diff
@@ -13,8 +13,9 @@ Canarytokens are place in CCCD in the following places;
 
 | Type | Memo | Notes |
 |---|---|---|
-| Cloned website | Cloned website detector on claim-crown-court-defence.service.gov.uk | Triggered when CCCD is viewed from an invalid domain.
-| MI Report | Fake reports access details file on '_environment_' - canary_base.docx | A report accessed via a hidden link on the MI report page. See below.
+| Cloned website | Cloned website detector on claim-crown-court-defence.service.gov.uk | Triggered when CCCD is viewed from an invalid domain. |
+| MI Report | Fake reports access details file on '_environment_' - canary_base.docx | A report accessed via a hidden link on the MI report page. [See below.](#mi-report) |
+| Dummy files on S3 | _filename_ in the S3 bucket for the '_environment_' environment - canary_base.docx | `tmp/contents.pdf` and `admin-users.docx` are Canarytokens placed on the S3 buckets for each of the environments. [See below.](#dummy-files-on-s3) |
```
Comment (jsugarman) [inline docs/thinkst_canary.md]: 🤔 should we be giving this info away in the repo

**Comment by jsugarman** [comment]
Also, just tried out and locally it should probably raise a useful message instead of failing with connection errors

File: spec/lib/thinkst_canary/token/shared_examples.rb (hunk 1 above)
Comment (jrmhaig) [reply spec/lib/thinkst_canary/token/shared_examples.rb:40]: `super()` is the previous version of `token_options`, from line 8.

**Review by jsugarman — APPROVED** [review]
👍

**Pull request merged**